File: returned/python/today11/single_number_ii.py
# https://leetcode.com/problems/single-number-ii/description/
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        ones = 0
        twos = 0

        for n in nums:
            ones = (ones ^ n) & (~twos)
            twos = (twos ^ n) & (~ones)

        return ones

File: typescript/today2/single-number-ii.ts
/**
 * https://leetcode.com/problems/single-number-ii/description/
 */
export { singleNumber };

function singleNumber(nums: number[]): number {
  let result = 0;

  for (let digitIndex = 0; digitIndex <= 31; digitIndex++) {
    const pow2Num = 1 << digitIndex;
    let digitSum = 0;
    for (let i = 0; i < nums.length; i++) {
      digitSum += (pow2Num & nums[i]) >> digitIndex;
    }
    const setBit = digitSum % 3;
    result += setBit << digitIndex;
  }

  return result;
}
